merge MyVirtualDirectory operator into operatorhubs

mlbiam opened the pull request:
created for both operatorhub.io and for the operatorhub in ocp4.  tested on ocp4.

SamiSousa commented on community-operators/myvirtualdirectory/myvirtualdirectory-operator.package.yaml:
```diff
@@ -0,0 +1,4 @@
+packageName: myvirtualdirectory-operator
+channels:
+- name: beta
+  currentCSV: myvirtualdirectory.1.0.0
```
nit: Please add newline

mlbiam replied:
A new line?  after line 4?  So there's a blank line 5?

SamiSousa commented on community-operators/myvirtualdirectory/myvirtualdirectory-operator.package.yaml:
```diff
@@ -0,0 +1,4 @@
+packageName: myvirtualdirectory-operator
```
This should match the operator's directory name, ie. `myvirtualdirectory`
Or the directory can be changed to match the packageName here.

SamiSousa commented on community-operators/myvirtualdirectory/myvirtualdirectory.crd.yaml:
```diff
@@ -0,0 +1,152 @@
+                                  cert_name:
+                                    type: string
+                                  key_name:
+                                    type: string
```
nit: Please add newline

mlbiam commented:
@SamiSousa thanks for the quick review.  updates made

SamiSousa commented on upstream-community-operators/myvirtualdirectory/myvirtualdirectory.1.0.0.clusterserviceversion.yaml:
```diff
@@ -0,0 +1,241 @@
+kind: ClusterServiceVersion
+metadata:
+  name: myvirtualdirectory.1.0.0
+  annotations:
```
Please add the `metadata.annotations.capabilities` field following the maturity model that can be found at this link: https://www.operatorhub.io/getting-started#How-do-I-start-writing-an-Operator?
Please apply this change to both CSVs.

mlbiam replied:
Done

SamiSousa commented on upstream-community-operators/myvirtualdirectory/myvirtualdirectory.1.0.0.clusterserviceversion.yaml:
```diff
@@ -0,0 +1,241 @@
+    certified: "false"
+    containerImage: docker.io/tremolosecurity/myvd-k8s-operator
+    createdAt: 2019-04-15
+    support: https://github.com/TremoloSecurity/myvd-k8s-operator
```
Please place the github repository into the `metadata.annotations.repository` field

SamiSousa commented:
@mlbiam Please address the comment above regarding the operator's git repository.
Also, it appears that scorecard is passing, but CRs aren't updated with a status section: https://travis-ci.com/operator-framework/community-operators/builds/108416179#L1704
This could mean that the CustomResource is created but not being watched by the operator. This will need to be manually verified

mlbiam commented:
@SamiSousa fixed the repository. Can you please explain what 

> Also, it appears that scorecard is passing, but CRs aren't updated with a status section

means?  We are definitely watching the CR that gets created

SamiSousa commented:
@mlbiam I don't doubt that. It's more likely that the operator wasn't picked up by OLM properly. This usually involves a step with using catalogsources to package the operator. I will test this by using OLM and operator-registry to run the scorecard test locally

SamiSousa commented:
@mlbiam The scorecard doesn't create a source secret, so the CR cannot be resolved. The operator gives the following error in this case:
```
in js : {"type":"ADDED","object":{"apiVersion":"myvirtualdirectory.tremolo.io/v1","kind":"MyVirtualDirectory","metadata":{"annotations":{"kubectl.kubernetes.io/last-applied-configuration":"{\"apiVersion\":\"myvirtualdirectory.tremolo.io/v1\",\"kind\":\"MyVirtualDirectory\",\"metadata\":{\"annotations\":{},\"name\":\"myvirtualdirectory\",\"namespace\":\"default\"},\"spec\":{\"dest_cfg_map\":\"myvd-configmap\",\"dest_secret\":\"myvd-secret\",\"image\":\"docker.io/tremolosecurity/myvirtualdirectory:latest\",\"key_store\":{\"key_pairs\":{\"create_keypair_template\":[{\"name\":\"ou\",\"value\":\"k8s\"},{\"name\":\"o\",\"value\":\"Tremolo Security\"},{\"name\":\"l\",\"value\":\"Alexandria\"},{\"name\":\"st\",\"value\":\"Virginia\"},{\"name\":\"c\",\"value\":\"US\"}],\"keys\":[{\"create_data\":{\"ca_cert\":false,\"key_size\":2048,\"server_name\":\"${k8s_obj.metadata.name + '.' + k8s_namespace + '.svc.cluster.local'}\",\"sign_by_k8s_ca\":false,\"subject_alternative_names\":[]},\"import_into_ks\":\"keypair\",\"name\":\"myvd-tls\",\"tls_secret_name\":\"myvd-tls-secret\"}]},\"static_keys\":[],\"trusted_certificates\":[]},\"myvd_log4j2_xml\":\"\\u003c?xml version=\\\"1.0\\\" encoding=\\\"UTF-8\\\"?\\u003e\\n\\u003c!--\\n    Copyright 2018 Tremolo Security, Inc.\\n    \\n    Licensed under the Apache License, Version 2.0 (the \\\"License\\\");\\n    you may not use this file except in compliance with the License.\\n    You may obtain a copy of the License at\\n    \\n        http://www.apache.org/licenses/LICENSE-2.0\\n    \\n    Unless required by applicable law or agreed to in writing, software\\n    distributed under the License is distributed on an \\\"AS IS\\\" BASIS,\\n    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.\\n    See the License for the specific language governing permissions and\\n    limitations under the License.\\n--\\u003e\\n\\n\\u003cConfiguration\\u003e\\n    \\u003cAppenders\\u003e\\n    \\u003cConsole name=\\\"STDOUT\\\" target=\\\"SYSTEM_OUT\\\"\\u003e\\n        \\u003cPatternLayout pattern=\\\"[%d][%t] %-5p %c{1} - %m%n\\\"/\\u003e\\n    \\u003c/Console\\u003e\\n    \\u003c/Appenders\\u003e\\n    \\u003cLoggers\\u003e\\n    \\n    \\u003cRoot level=\\\"info\\\"\\u003e\\n        \\u003cAppenderRef ref=\\\"STDOUT\\\"/\\u003e\\n    \\u003c/Root\\u003e\\n    \\u003c/Loggers\\u003e\\n\\u003c/Configuration\\u003e\",\"myvd_network_configuration\":{\"allowed_client_names\":[],\"ciphers\":[\"TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256\",\"TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256\",\"TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA\",\"TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384\",\"TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA384\",\"TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA\"],\"client_auth\":\"none\",\"open_port\":10389,\"path_to_deployment\":\"/etc/myvd-config\",\"path_to_env_file\":\"/etc/myvd\",\"secure_key_alias\":\"myvd-tls\",\"secure_port\":10636},\"myvd_props\":\"#Global AuthMechConfig\\nserver.globalChain=accesslog,attrCleaner\\nserver.globalChain.accesslog.className=com.tremolosecurity.proxy.myvd.log.AccessLog\\nserver.globalChain.attrCleaner.className=net.sourceforge.myvd.inserts.mapping.AttributeCleaner\\nserver.globalChain.attrCleaner.config.clearAttributes=true\\nserver.globalChain.dump.className=net.sourceforge.myvd.inserts.DumpTransaction\\nserver.globalChain.dump.config.logLevel=info\\nserver.globalChain.dump.config.label=global\\n\\nserver.nameSpaces=rootdse,myvdroot\\nserver.rootdse.chain=dse\\nserver.rootdse.nameSpace=\\nserver.rootdse.weight=0\\nserver.rootdse.dse.className=net.sourceforge.myvd.inserts.RootDSE\\nserver.rootdse.dse.config.namingContexts=#[ROOT]\\nserver.myvdroot.chain=root\\nserver.myvdroot.nameSpace=#[ROOT]\\nserver.myvdroot.weight=0\\nserver.myvdroot.root.className=net.sourceforge.myvd.inserts.RootObject\",\"non_secret_data\":[{\"name\":\"ROOT\",\"value\":\"o=Tremolo\"}],\"replicas\":1,\"secret_data\":[\"MYVD_TLS_KEY_PASSWORD\"],\"source_secret\":\"myvd-secrets-source\"}}\n"},"creationTimestamp":"2019-04-16T18:59:59Z","generation":1,"name":"myvirtualdirectory","namespace":"default","resourceVersion":"43317","selfLink":"/apis/myvirtualdirectory.tremolo.io/v1/namespaces/default/myvirtualdirectories/myvirtualdirectory","uid":"d5237499-6079-11e9-b91b-14fb74ee5482"},"spec":{"dest_cfg_map":"myvd-configmap","dest_secret":"myvd-secret","image":"docker.io/tremolosecurity/myvirtualdirectory:latest","key_store":{"key_pairs":{"create_keypair_template":[{"name":"ou","value":"k8s"},{"name":"o","value":"Tremolo Security"},{"name":"l","value":"Alexandria"},{"name":"st","value":"Virginia"},{"name":"c","value":"US"}],"keys":[{"create_data":{"ca_cert":false,"key_size":2048,"server_name":"${k8s_obj.metadata.name + '.' + k8s_namespace + '.svc.cluster.local'}","sign_by_k8s_ca":false,"subject_alternative_names":[]},"import_into_ks":"keypair","name":"myvd-tls","tls_secret_name":"myvd-tls-secret"}]},"static_keys":[],"trusted_certificates":[]},"myvd_log4j2_xml":"\u003c?xml version=\"1.0\" encoding=\"UTF-8\"?\u003e\n\u003c!--\n    Copyright 2018 Tremolo Security, Inc.\n    \n    Licensed under the Apache License, Version 2.0 (the \"License\");\n    you may not use this file except in compliance with the License.\n    You may obtain a copy of the License at\n    \n        http://www.apache.org/licenses/LICENSE-2.0\n    \n    Unless required by applicable law or agreed to in writing, software\n    distributed under the License is distributed on an \"AS IS\" BASIS,\n    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.\n    See the License for the specific language governing permissions and\n    limitations under the License.\n--\u003e\n\n\u003cConfiguration\u003e\n    \u003cAppenders\u003e\n    \u003cConsole name=\"STDOUT\" target=\"SYSTEM_OUT\"\u003e\n        \u003cPatternLayout pattern=\"[%d][%t] %-5p %c{1} - %m%n\"/\u003e\n    \u003c/Console\u003e\n    \u003c/Appenders\u003e\n    \u003cLoggers\u003e\n    \n    \u003cRoot level=\"info\"\u003e\n        \u003cAppenderRef ref=\"STDOUT\"/\u003e\n    \u003c/Root\u003e\n    \u003c/Loggers\u003e\n\u003c/Configuration\u003e","myvd_network_configuration":{"allowed_client_names":[],"ciphers":["TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256","TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256","TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA","TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384","TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA384","TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA"],"client_auth":"none","open_port":10389,"path_to_deployment":"/etc/myvd-config","path_to_env_file":"/etc/myvd","secure_key_alias":"myvd-tls","secure_port":10636},"myvd_props":"#Global AuthMechConfig\nserver.globalChain=accesslog,attrCleaner\nserver.globalChain.accesslog.className=com.tremolosecurity.proxy.myvd.log.AccessLog\nserver.globalChain.attrCleaner.className=net.sourceforge.myvd.inserts.mapping.AttributeCleaner\nserver.globalChain.attrCleaner.config.clearAttributes=true\nserver.globalChain.dump.className=net.sourceforge.myvd.inserts.DumpTransaction\nserver.globalChain.dump.config.logLevel=info\nserver.globalChain.dump.config.label=global\n\nserver.nameSpaces=rootdse,myvdroot\nserver.rootdse.chain=dse\nserver.rootdse.nameSpace=\nserver.rootdse.weight=0\nserver.rootdse.dse.className=net.sourceforge.myvd.inserts.RootDSE\nserver.rootdse.dse.config.namingContexts=#[ROOT]\nserver.myvdroot.chain=root\nserver.myvdroot.nameSpace=#[ROOT]\nserver.myvdroot.weight=0\nserver.myvdroot.root.className=net.sourceforge.myvd.inserts.RootObject","non_secret_data":[{"name":"ROOT","value":"o=Tremolo"}],"replicas":1,"secret_data":["MYVD_TLS_KEY_PASSWORD"],"source_secret":"myvd-secrets-source"}}}
Problem calling '/api/v1/namespaces/default/secrets/myvd-secrets-source' - 404
{"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"secrets \"myvd-secrets-source\" not found","reason":"NotFound","details":{"name":"myvd-secrets-source","kind":"secrets"},"code":404}

Sleeping, then trying again
trying again
Problem calling '/api/v1/namespaces/default/secrets/myvd-secrets-source' - 404
{"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"secrets \"myvd-secrets-source\" not found","reason":"NotFound","details":{"name":"myvd-secrets-source","kind":"secrets"},"code":404}
```
Eventually an application pod comes up but is unstable.
When I create the `myvd-secrets-source` secret, and delete the CR and reapply it, I get the following error from the operator:
```
Processing key 'myvd-tls'
Checking if kubernetes secret exists
Verification for '/api/v1/namespaces/default/secrets/myvd-tls-secret' failed / {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"secrets \"myvd-tls-secret\" not found","reason":"NotFound","details":{"name":"myvd-tls-secret","kind":"secrets"},"code":404}

Creating keypair
Creating secret
Posting secret
Storing to keystore
Exception in thread "main" java.lang.RuntimeException: java.security.KeyStoreException: Key protection  algorithm not found: java.security.UnrecoverableKeyException: Encrypt Private Key failed: password has been cleared
	at jdk.nashorn.internal.runtime.ScriptRuntime.apply(ScriptRuntime.java:397)
	at jdk.nashorn.api.scripting.ScriptObjectMirror.callMember(ScriptObjectMirror.java:199)
	at jdk.nashorn.api.scripting.NashornScriptEngine.invokeImpl(NashornScriptEngine.java:386)
	at jdk.nashorn.api.scripting.NashornScriptEngine.invokeFunction(NashornScriptEngine.java:190)
	at com.tremolosecurity.kubernetes.artifacts.util.K8sUtils.watchURI(K8sUtils.java:246)
	at com.tremolosecurity.kubernetes.artifacts.run.Controller.runWatch(Controller.java:160)
	at com.tremolosecurity.kubernetes.artifacts.run.Controller.main(Controller.java:121)
Caused by: java.security.KeyStoreException: Key protection  algorithm not found: java.security.UnrecoverableKeyException: Encrypt Private Key failed: password has been cleared
	at sun.security.pkcs12.PKCS12KeyStore.setKeyEntry(PKCS12KeyStore.java:674)
	at sun.security.pkcs12.PKCS12KeyStore.engineSetKeyEntry(PKCS12KeyStore.java:574)
	at java.security.KeyStore.setKeyEntry(KeyStore.java:1140)
	at com.tremolosecurity.kubernetes.artifacts.util.CertUtils.saveX509ToKeystore(CertUtils.java:363)
	at jdk.nashorn.internal.scripts.Script$Recompilation$65$1472A$\^eval\_.process_key_pair_config(<eval>:245)
	at jdk.nashorn.internal.scripts.Script$Recompilation$61$12413A$\^eval\_.generate_myvd_secret(<eval>:379)
	at jdk.nashorn.internal.scripts.Script$Recompilation$59$52A$\^eval\_.on_watch(<eval>:9)
	at jdk.nashorn.internal.runtime.ScriptFunctionData.invoke(ScriptFunctionData.java:639)
	at jdk.nashorn.internal.runtime.ScriptFunction.invoke(ScriptFunction.java:494)
	at jdk.nashorn.internal.runtime.ScriptRuntime.apply(ScriptRuntime.java:393)
	... 6 more
Caused by: java.security.UnrecoverableKeyException: Encrypt Private Key failed: password has been cleared
	at sun.security.pkcs12.PKCS12KeyStore.encryptPrivateKey(PKCS12KeyStore.java:918)
	at sun.security.pkcs12.PKCS12KeyStore.setKeyEntry(PKCS12KeyStore.java:611)
	... 15 more
Caused by: java.lang.IllegalStateException: password has been cleared
	at javax.crypto.spec.PBEKeySpec.getPassword(PBEKeySpec.java:197)
	at org.bouncycastle.jcajce.provider.symmetric.util.BCPBEKey.getPassword(Unknown Source)
	at com.sun.crypto.provider.PKCS12PBECipherCore.implInit(PKCS12PBECipherCore.java:264)
	at com.sun.crypto.provider.PKCS12PBECipherCore.implInit(PKCS12PBECipherCore.java:370)
	at com.sun.crypto.provider.PKCS12PBECipherCore.implInit(PKCS12PBECipherCore.java:355)
	at com.sun.crypto.provider.PKCS12PBECipherCore$PBEWithSHA1AndDESede.engineInit(PKCS12PBECipherCore.java:464)
	at javax.crypto.Cipher.implInit(Cipher.java:809)
	at javax.crypto.Cipher.chooseProvider(Cipher.java:863)
	at javax.crypto.Cipher.init(Cipher.java:1538)
	at javax.crypto.Cipher.init(Cipher.java:1469)
	at sun.security.pkcs12.PKCS12KeyStore.encryptPrivateKey(PKCS12KeyStore.java:901)
	... 16 more
Cought the shutdown hook
```
Is this a known issue?
I'm unsure how to test the operator since the description does not describe what the source secret should look like, and does not mention the other secret and configmap named in the example CR.

mlbiam commented:
Try reloading the operator.  I added some docs to "Before You Start" to tell you how to create the source secret.  I also added some better error management code to the operator so if the secret doesn't exist it will fail more gracefully and with a better error message.

SamiSousa commented:
Following the new instructions, I'm able to create a secret that gets picked up by the operator and results in a deployment and service being created. Looks good!

SamiSousa commented:
/lgtm
